### DISCUSSION TITLE
Stitching photographs of a mosaic with black seams

### DISCUSSION THREAD

**Discussion by mstenta** [OpenStitching/stitching#254] [Q&A]
I am looking for advice on how to optimize seam finding for a specific use case. I am photographing a piece of mosaic artwork with a digital camera on a copy stand. The mosaic is comprised of tile pieces separated by pure black lines. The `AffineStitcher` works really well at stitching the photographs together, but the seam lines often go through the middle of the mosaic tile pieces, resulting in visible seams upon close inspection. If it were possible for the seam lines to follow the black lines in between the tile pieces, then it would result in a perfect final stitch without any visible seams. Is there any way to achieve this?

**Comment by mstenta** [comment]
I have tested all of the different seam finder choices, and so far I have found that `gc_color` comes closest to following the black seams (although not always perfectly).

Another problem, however, seems to be related to the "block size" of the seam lines (if that is the correct terminology).

Attached below are some example close-ups of what I'm experiencing. Notice how the seam line almost follows the black area between the colorful areas, but it cuts across the colorful areas in some places, which results in visible artifacts.

The first shows a close-up with seam line visible:

![Screenshot from 2024-11-11 13-52-49](https://github.com/user-attachments/assets/7807b257-7871-4752-91a6-178dcb0288de)

Here is the same close-up without the seam line, so you can see the artifacts along the edge:

![Screenshot from 2024-11-11 13-52-56](https://github.com/user-attachments/assets/6e70a640-afe1-4453-bcac-d24f9cdaf3e2)

And here is the original photograph of the same area, showing what it should look like:

![Screenshot from 2024-11-11 13-53-01](https://github.com/user-attachments/assets/5d3eb1ee-8d03-4e57-bfc5-5b2569ba487f)

Is it possible to adjust the seam finding logic such that:

a) it always follows the black lines

and/or:

b) it uses smaller "block size" so that the path of the seam is less "chunky"

(a) would be the ideal solution, but if that isn't possible then (b) may help at least.

**Comment by mstenta** [comment]
> smaller "block size" so that the path of the seam is less "chunky"

Ah I understand why this happens now. It is related to the fact that the images are downsized to a lower resolution before the seam finding is performed, and then resized to the final resolution. So the seam lines become "course" when they are enlarged to fit the final image.

If I set `low_megapix` and `medium_megapix` both to `-1` (full resolution), it seems to work quite well! Although it takes a lot longer... but that isn't surprising. :sweat_smile:

Here is the same seam line using full resolution images:

![Screenshot from 2024-11-11 14-09-36](https://github.com/user-attachments/assets/2e39fc0c-29ce-40df-9733-68a4d80f7e7b)

**Reply by lukasalexanderweber** [reply]
great! I've also stumbled across the "block size" in seam lines. as you say, it's a tradeoff between accuracy and performance. Do you still have areas that are cuted? 

I have wanted to introduce object-based seam detection for some time now. This would allow users to mark elements through which no seams should pass. There is some research [here](https://www.ecva.net/papers/eccv_2018/papers_ECCV/papers/Charles_Herrmann_Object-centered_image_stitching_ECCV_2018_paper.pdf). As far as I understand it, the algorithms use an energy map and the seam then goes along areas with little energy ([1](https://github.com/andrewdcampbell/seam-carving), [2](https://github.com/vivianhylee/seam-carving)). So if you don't want the seam to go through an area, you can simply increase the energy there. For experimental purposes, I simply placed the [starry night](https://en.wikipedia.org/wiki/The_Starry_Night) over the areas I didn't want seams to pass through, and that worked well. In the long run it would be nice to implement that, but I guess we don't have python acess on the c++ energy map.

**Reply by mstenta** [reply]
Object-based seam detection... Interesting!

> Do you still have areas that are cuted?

Not in this specific case... it worked well! I have others that I will be stitching in the future, so I will see if `gc_color` is consistently my best choice, or if different images require different approaches... to be continued!

Thanks for the quick reply and ideas!
